From cd0a44f5ff81e2c3b666d4f65b3d06a692646c10 Mon Sep 17 00:00:00 2001
From: Aaradhana Sahu <aaradhana.sahu@oss.qualcomm.com>
Date: Thu, 24 Jul 2025 12:33:02 +0530
Subject: [PATCH] QSDK: DP: Fix WARN_ON during SSR on qcn6432

During DP interrupt registration, netif_napi_add_weight() is called with
ATH12K_EXT_IRQ_DP_NUM_VECTORS (16), but only ATH12K_EXT_IRQ_GRP_NUM_MAX
(12) interrupts are freed. As a result, during recovery,
netif_napi_add_weight() triggers a warning because some NAPI contexts
still have the NAPI_STATE_LISTED flag set.

call tarce:

netif_napi_add_weight+0x1b0/0x20c
ath12k_pcic_get_msi_data+0x32c/0x5c0 [ath12k]
ath12k_pcic_cfg_hybrid_ext_irq+0x11c/0x14c [ath12k]
ath12k_wifi7_sdwf_reinject_handler+0x7c/0x508 [ath12k_wifi7]
ath12k_dp_cmn_device_init+0x40/0x7c4 [ath12k]
ath12k_core_qmi_firmware_ready+0x134/0xdf0 [ath12k]
ath12k_core_qmi_firmware_ready+0xb34/0xdf0 [ath12k]
process_scheduled_works+0x1fc/0x2f4
worker_thread+0x218/0x284
kthread+0x10c/0x11c
ret_from_fork+0x10/0x20

Fix this issue by checking ATH12K_EXT_IRQ_GRP_NUM_MAX instead of
ATH12K_EXT_IRQ_DP_NUM_VECTORS for qcn6432 during DP interrupt registration.

Patch-dependency: None
Patch-work: None

Fixes: I8ed64ec1ed8 ("QSDK: DP wifi: ath12k: Tx KPI Optimizations")
Signed-off-by: Aaradhana Sahu <aaradhana.sahu@oss.qualcomm.com>
---
 drivers/net/wireless/ath/ath12k/pcic.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/net/wireless/ath/ath12k/pcic.c b/drivers/net/wireless/ath/ath12k/pcic.c
index 9592584..6dfd9bd 100644
--- a/drivers/net/wireless/ath/ath12k/pcic.c
+++ b/drivers/net/wireless/ath/ath12k/pcic.c
@@ -1069,7 +1069,7 @@ int ath12k_pcic_msi_desc_assign_irq(struct ath12k_base *ab,
 			continue;
 		}
 
-		if (i >= (base_vector + num_vectors) || *k >= ATH12K_EXT_IRQ_DP_NUM_VECTORS)
+		if (i >= (base_vector + num_vectors) || *k >= ATH12K_EXT_IRQ_GRP_NUM_MAX)
 			break;
 
 		ret = ath12k_pcic_ext_cfg_gic_msi_irq(ab, irq_handler, dp,
-- 
2.34.1

